src/data/sum/basic.lean

```diff
@@ -348,3 +348,20 @@ lemma surjective.sum_map {f : α → β} {g : α' → β'} (hf : surjective f) (
 | (inr y) := let ⟨x, hx⟩ := hg y in ⟨inr x, congr_arg inr hx⟩
 
 end function
+
+/-!
+### Ternary sum
+
+Abbreviations for the maps from the summands to `α ⊕ β ⊕ γ`. This is useful for pattern-matching.
+-/
+
+namespace sum3
+
+/-- The map from the first summand into a ternary sum. -/
+@[pattern, simp, reducible] def in₀ (a) : α ⊕ β ⊕ γ := inl a
```

Comment:
I assume something ridiculous like
```lean
class sum.has_in (α : out_param Type) (β) (n : ℕ) :=
(in_ : α → β)

instance : has_in α (α ⊕ β) 0 := ⟨sum.inl⟩
instance [has_in γ β n] : has_in γ (α ⊕ β) (succ n) := ⟨ λ c, sum.inr (in_ c)⟩
```
Won't work here?

Reply:
Oh goodness no, please 🤣

I unironically thought about this, but the problem is that `in_ 2` could refer both to `inr` and `inr ∘ inl`, `in_ 3` could refer both to `inr ∘ inr` and `inr ∘ inr ∘ inl`, etc... You will note that your snippet above doesn't allow this because `in_` won't ever be only a bunch of `inr`.

Reply:
I played with this a little more and wasn't quite able to get it to work